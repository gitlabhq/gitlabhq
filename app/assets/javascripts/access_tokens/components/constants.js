import { __, s__ } from '~/locale';

export const EVENT_ERROR = 'ajax:error';
export const EVENT_SUCCESS = 'ajax:success';
export const FORM_SELECTOR = '#js-new-access-token-form';

export const INITIAL_PAGE = 1;
export const PAGE_SIZE = 100;

const BASE_FIELDS = [
  {
    key: 'name',
    label: __('Token name'),
    isRowHeader: true,
    sortable: true,
  },
  {
    formatter(description) {
      return description ?? '-';
    },
    key: 'description',
    label: __('Description'),
    sortable: true,
  },
  {
    formatter(scopes) {
      return scopes?.length ? scopes.join(', ') : __('no scopes selected');
    },
    key: 'scopes',
    label: __('Scopes'),
    sortable: true,
  },
  {
    key: 'createdAt',
    label: s__('AccessTokens|Created'),
    sortable: true,
  },
  {
    key: 'lastUsedAt',
    label: __('Last Used'),
    sortable: true,
  },
];

const ROLE_FIELD = {
  key: 'role',
  label: __('Role'),
  sortable: true,
};

export const FIELDS = [
  ...BASE_FIELDS,
  {
    formatter(ips) {
      return ips?.length ? ips?.join(', ') : '-';
    },
    key: 'lastUsedIps',
    label: __('Last Used IPs'),
    sortable: false,
    tdClass: 'gl-break-all gl-min-w-20',
  },
  {
    key: 'expiresAt',
    label: __('Expires'),
    sortable: true,
  },
  ROLE_FIELD,
  {
    key: 'action',
    label: __('Action'),
    tdClass: '!gl-py-3',
  },
];

export const INACTIVE_TOKENS_TABLE_FIELDS = [
  ...BASE_FIELDS,
  {
    key: 'expiresAt',
    label: __('Expired'),
    sortable: true,
  },
  ROLE_FIELD,
];
