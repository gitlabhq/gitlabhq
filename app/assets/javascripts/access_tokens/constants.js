import { GlFilteredSearchToken } from '@gitlab/ui';
import { s__, __ } from '~/locale';
import {
  OPERATORS_AFTER_BEFORE,
  OPERATORS_IS,
} from '~/vue_shared/components/filtered_search_bar/constants';
import DateToken from '~/vue_shared/components/filtered_search_bar/tokens/date_token.vue';

// Token types
export const FEED_TOKEN = 'feedToken';
export const INCOMING_EMAIL_TOKEN = 'incomingEmailToken';
export const STATIC_OBJECT_TOKEN = 'staticObjectToken';

export const DEFAULT_FILTER = [
  {
    type: 'state',
    value: {
      data: 'active',
      operator: '=',
    },
  },
];

export const FILTER_OPTIONS = [
  {
    icon: 'status',
    title: s__('AccessTokens|State'),
    type: 'state',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'active', title: s__('AccessTokens|Active') },
      { value: 'inactive', title: s__('AccessTokens|Inactive') },
    ],
  },
  {
    icon: 'remove',
    title: s__('AccessTokens|Revoked'),
    type: 'revoked',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'true', title: __('Yes') },
      { value: 'false', title: __('No') },
    ],
  },
  {
    icon: 'history',
    title: __('Created date'),
    type: 'created',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Expiration date'),
    type: 'expires',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Last used date'),
    type: 'last_used',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
];

export const FILTER_OPTIONS_CREDENTIALS_INVENTORY = [
  {
    icon: 'key',
    title: s__('CredentialsInventory|Type'),
    type: 'filter',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      {
        value: 'personal_access_tokens',
        title: s__('CredentialsInventory|Personal access tokens'),
      },
      { value: 'ssh_keys', title: s__('CredentialsInventory|SSH keys') },
      {
        value: 'resource_access_tokens',
        title: s__('CredentialsInventory|Project and group access tokens'),
      },
      { value: 'gpg_keys', title: s__('CredentialsInventory|GPG keys') },
    ],
  },
  {
    icon: 'user',
    title: s__('CredentialsInventory|Owner type'),
    type: 'owner_type',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'human', title: __('Human') },
      { value: 'service_account', title: __('Service account') },
    ],
  },
  ...FILTER_OPTIONS,
];

export const DEFAULT_SORT = { value: 'expires', isAsc: true };

export const SORT_OPTIONS = [
  {
    text: __('Created date'),
    value: 'created',
    sort: {
      asc: 'created_asc',
      desc: 'created_desc',
    },
  },
  {
    text: __('Expiration date'),
    value: 'expires',
    sort: {
      asc: 'expires_asc',
      desc: 'expires_desc',
    },
  },
  {
    text: __('Last used date'),
    value: 'last_used',
    sort: {
      asc: 'last_used_asc',
      desc: 'last_used_desc',
    },
  },
  {
    text: __('Name'),
    value: 'name',
    sort: {
      asc: 'name_asc',
      desc: 'name_desc',
    },
  },
];

export const STATISTICS_CONFIG = [
  {
    title: s__('AccessTokens|Active tokens'),
    tooltipTitle: s__('AccessTokens|Filter for active tokens'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Tokens expiring in 2 weeks'),
    tooltipTitle: s__('AccessTokens|Filter for tokens expiring in 2 weeks'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
      {
        type: 'expires',
        value: {
          data: 'DATE_HOLDER',
          operator: '<',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Revoked tokens'),
    tooltipTitle: s__('AccessTokens|Filter for revoked tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'true',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Expired tokens'),
    tooltipTitle: s__('AccessTokens|Filter for expired tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'false',
          operator: '=',
        },
      },
      {
        type: 'state',
        value: {
          data: 'inactive',
          operator: '=',
        },
      },
    ],
  },
];
