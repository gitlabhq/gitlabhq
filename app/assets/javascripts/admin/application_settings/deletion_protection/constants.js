import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  label: s__('DeletionSettings|Retention period'),
  helpText: s__(
    'DeletionSettings|Period that deleted groups and projects will remain restorable for.',
  ),
  learnMore: __('Learn more.'),
  days: __('days'),
};

export const DEL_ADJ_PERIOD_MAX_LIMIT = 90;
export const DEL_ADJ_PERIOD_MIN_LIMIT = 1;

export const DEL_ADJ_PERIOD_MAX_LIMIT_ERROR = s__(
  'DeletionSettings|Maximum deletion protection duration is 90 days.',
);

export const DEL_ADJ_PERIOD_MIN_LIMIT_ERROR = s__(
  'DeletionSettings|Minimum deletion protection duration is 1 day.',
);
