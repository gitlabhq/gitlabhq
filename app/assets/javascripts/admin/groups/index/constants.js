import groupsEmptyStateIllustration from '@gitlab/svgs/dist/illustrations/empty-state/empty-groups-md.svg?url';
import { s__, __ } from '~/locale';
import {
  SORT_LABEL_NAME,
  SORT_LABEL_CREATED,
  SORT_LABEL_UPDATED,
  SORT_LABEL_STORAGE_SIZE,
} from '~/groups_projects/constants';
import GroupsList from '~/vue_shared/components/groups_list/groups_list.vue';
import { formatGraphQLGroups } from '~/vue_shared/components/groups_list/formatter';
import ResourceListsEmptyState from '~/vue_shared/components/resource_lists/empty_state.vue';
import adminGroupsQuery from './graphql/queries/groups.query.graphql';

const baseTab = {
  formatter: (groups) =>
    formatGraphQLGroups(groups, (group) => {
      const adminPath = `/admin/groups/${group.fullPath}`;

      return {
        avatarLabelLink: adminPath,
        editPath: `${adminPath}/edit`,
      };
    }),
  listComponent: GroupsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showGroupIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  query: adminGroupsQuery,
  queryPath: 'groups',
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  variables: { active: true },
  countsQueryPath: 'active',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any active groups yet."),
    description: s__(
      'Organization|A group is a collection of several projects. If you organize your projects under a group, it works like a folder.',
    ),
    'data-testid': 'groups-empty-state',
  },
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  variables: { active: false },
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any inactive groups."),
    description: s__('Groups|Groups that are archived or pending deletion will appear here.'),
  },
};

export const SORT_OPTION_NAME = {
  value: 'name',
  text: SORT_LABEL_NAME,
};

export const SORT_OPTION_CREATED = {
  value: 'created_at',
  text: SORT_LABEL_CREATED,
};

export const SORT_OPTION_UPDATED = {
  value: 'updated_at',
  text: SORT_LABEL_UPDATED,
};

export const SORT_OPTION_STORAGE_SIZE = {
  value: 'storage_size',
  text: SORT_LABEL_STORAGE_SIZE,
};

export const SORT_OPTIONS = [
  SORT_OPTION_NAME,
  SORT_OPTION_CREATED,
  SORT_OPTION_UPDATED,
  SORT_OPTION_STORAGE_SIZE,
];

export const ADMIN_GROUPS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/groups';

export const ADMIN_GROUPS_ROUTE_NAME = 'admin-groups';

export const FIRST_TAB_ROUTE_NAMES = [ADMIN_GROUPS_ROUTE_NAME];

export const FILTERED_SEARCH_TERM_KEY = 'search';
export const FILTERED_SEARCH_NAMESPACE = 'admin-groups';
