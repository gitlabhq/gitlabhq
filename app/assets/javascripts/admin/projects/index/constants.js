import emptyStateProjectsSvgPath from '@gitlab/svgs/dist/illustrations/empty-state/empty-projects-md.svg?url';
import { __, s__ } from '~/locale';
import ProjectsList from '~/vue_shared/components/projects_list/projects_list.vue';
import ResourceListsEmptyState, {
  TYPES,
} from '~/vue_shared/components/resource_lists/empty_state.vue';
import { formatGraphQLProjects } from '~/vue_shared/components/projects_list/formatter';
import projectsQuery from './graphql/queries/projects.query.graphql';

const baseTab = {
  listComponent: ProjectsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showProjectIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  emptyStateComponentProps: {
    title: s__('Projects|No projects found.'),
    svgPath: emptyStateProjectsSvgPath,
    searchMinimumLength: 3,
    type: TYPES.filter,
  },
  queryPath: 'projects',
  formatter: (projects) =>
    formatGraphQLProjects(projects, (project) => ({
      editPath: `/admin/projects/${project.fullPath}/edit`,
      avatarLabelLink: `/admin/projects/${project.fullPath}`,
    })),
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  query: projectsQuery,
  variables: { active: true },
  queryPath: 'projects',
  countsQueryPath: 'active',
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  query: projectsQuery,
  variables: { active: false },
  queryPath: 'projects',
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    ...baseTab.emptyStateComponentProps,
    title: s__("Projects|You don't have any inactive projects."),
    description: s__('Projects|Projects that are archived or pending deletion will appear here.'),
  },
};

export const ADMIN_PROJECTS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/projects';

export const ADMIN_PROJECTS_ROUTE_NAME = 'admin-projects';
export const FIRST_TAB_ROUTE_NAMES = [ADMIN_PROJECTS_ROUTE_NAME];

export const FILTERED_SEARCH_NAMESPACE = 'admin';
export const FILTERED_SEARCH_TERM_KEY = 'search';
