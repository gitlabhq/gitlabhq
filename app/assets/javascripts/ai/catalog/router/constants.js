export const AI_CATALOG_INDEX_ROUTE = 'ai-catalog';
