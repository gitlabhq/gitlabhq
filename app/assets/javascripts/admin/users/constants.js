import { GlFilteredSearchToken } from '@gitlab/ui';

import { OPERATORS_IS } from '~/vue_shared/components/filtered_search_bar/constants';
import { s__, __, sprintf } from '~/locale';

export const I18N_USER_ACTIONS = {
  edit: __('Edit'),
  editWithName: (name) => sprintf(__('Edit user: %{user_name}'), { user_name: name }),
  userAdministration: s__('AdminUsers|User administration'),
  unlock: __('Unlock'),
  block: s__('AdminUsers|Block'),
  unblock: s__('AdminUsers|Unblock'),
  approve: s__('AdminUsers|Approve'),
  reject: s__('AdminUsers|Reject'),
  deactivate: s__('AdminUsers|Deactivate'),
  activate: s__('AdminUsers|Activate'),
  ldapBlocked: s__('AdminUsers|Cannot unblock LDAP blocked users'),
  delete: s__('AdminUsers|Delete user'),
  deleteWithContributions: s__('AdminUsers|Delete user and contributions'),
  ban: s__('AdminUsers|Ban user'),
  unban: s__('AdminUsers|Unban user'),
  trust: s__('AdminUsers|Trust user'),
  untrust: s__('AdminUsers|Untrust user'),
};

// NOTE: the 'access_level' token is overridden in EE version
export const ACCESS_LEVEL_TOKEN_TYPE = 'access_level';

export const TOKENS = [
  {
    title: s__('AdminUsers|Access level'),
    type: ACCESS_LEVEL_TOKEN_TYPE,
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'admins', title: s__('AdminUsers|Administrator') },
      { value: 'external', title: s__('AdminUsers|External') },
    ],
  },
  {
    title: __('State'),
    type: 'state',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'active', title: s__('AdminUsers|Active') },
      { value: 'banned', title: s__('AdminUsers|Banned') },
      { value: 'blocked', title: s__('AdminUsers|Blocked') },
      { value: 'deactivated', title: s__('AdminUsers|Deactivated') },
      {
        value: 'blocked_pending_approval',
        title: s__('AdminUsers|Pending approval'),
      },
      { value: 'trusted', title: s__('AdminUsers|Trusted') },
      { value: 'wop', title: s__('AdminUsers|Without projects') },
    ],
  },
  {
    title: s__('AdminUsers|Two-factor authentication'),
    type: '2fa',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'two_factor_enabled', title: __('On') },
      { value: 'two_factor_disabled', title: __('Off') },
    ],
  },
  {
    title: __('Type'),
    type: 'type',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'without_bots', title: s__('AdminUsers|Humans') },
      { value: 'bots', title: s__('AdminUsers|Bots') },
      { value: 'placeholder', title: s__('UserMapping|Placeholder') },
    ],
  },
  {
    title: s__('AdminUsers|LDAP sync'),
    type: 'ldap_sync',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [{ value: 'ldap_sync', title: __('True') }],
  },
];

export const SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT = 10;

export const SOLO_OWNED_ORGANIZATIONS_EMPTY = {
  count: 0,
  nodes: [],
};
