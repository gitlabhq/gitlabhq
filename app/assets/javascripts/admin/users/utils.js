import { queryToObject } from '~/lib/utils/url_utility';
import getSoloOwnedOrganizationsQuery from '~/admin/users/graphql/queries/get_solo_owned_organizations.query.graphql';
import { convertToGraphQLId } from '~/graphql_shared/utils';
import { TYPENAME_USER } from '~/graphql_shared/constants';
import {
  TOKENS,
  SOLO_OWNED_ORGANIZATIONS_EMPTY,
  SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
} from 'ee_else_ce/admin/users/constants';

export const generateUserPaths = (paths, id) => {
  return Object.fromEntries(
    Object.entries(paths).map(([action, genericPath]) => {
      return [action, genericPath.replace('id', id)];
    }),
  );
};

/**
 * @typedef {{type: string, value: {data: string, operator: string}}} Token
 */

/**
 * Initialize token values based on the URL parameters
 * @param {string} query - document.location.searchd
 *
 * @returns {{tokens: Array<string|Token>, sort: string}}
 */
export function initializeValuesFromQuery(query = document.location.search) {
  const tokens = [];

  const { filter, search_query: searchQuery, sort } = queryToObject(query);

  if (filter) {
    const token = TOKENS.find(({ options }) => options.some(({ value }) => value === filter));

    if (token) {
      tokens.push({
        type: token.type,
        value: {
          data: filter,
          operator: token.operators[0].value,
        },
      });
    }
  }

  if (searchQuery) {
    tokens.push(searchQuery);
  }

  return { tokens, sort };
}

export const getSoloOwnedOrganizations = async (apolloClient, userId) => {
  if (!window.gon?.features?.uiForOrganizations) {
    return Promise.resolve(SOLO_OWNED_ORGANIZATIONS_EMPTY);
  }

  const {
    data: {
      user: { organizations },
    },
  } = await apolloClient.query({
    query: getSoloOwnedOrganizationsQuery,
    variables: {
      id: convertToGraphQLId(TYPENAME_USER, userId),
      first: SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
    },
  });

  return organizations;
};
