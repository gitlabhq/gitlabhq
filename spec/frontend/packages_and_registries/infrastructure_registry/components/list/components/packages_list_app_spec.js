import { GlEmptyState, GlSprintf, GlLink } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
// eslint-disable-next-line no-restricted-imports
import Vuex from 'vuex';
import setWindowLocation from 'helpers/set_window_location_helper';
import { helpPagePath } from '~/helpers/help_page_helper';
import { createAlert, VARIANT_INFO } from '~/alert';
import * as commonUtils from '~/lib/utils/common_utils';
import PackageListApp from '~/packages_and_registries/infrastructure_registry/list/components/packages_list_app.vue';
import { DELETE_PACKAGE_SUCCESS_MESSAGE } from '~/packages_and_registries/infrastructure_registry/list/constants';
import { SHOW_DELETE_SUCCESS_ALERT } from '~/packages_and_registries/shared/constants';

import * as packageUtils from '~/packages_and_registries/shared/utils';
import InfrastructureSearch from '~/packages_and_registries/infrastructure_registry/list/components/infrastructure_search.vue';
import InfrastructureTitle from '~/packages_and_registries/infrastructure_registry/list/components/infrastructure_title.vue';
import { FILTERED_SEARCH_TERM } from '~/vue_shared/components/filtered_search_bar/constants';

jest.mock('~/lib/utils/common_utils');
jest.mock('~/alert');

Vue.use(Vuex);

describe('packages_list_app', () => {
  let wrapper;
  let store;

  const PackageList = {
    name: 'package-list',
    template: '<div><slot name="empty-state"></slot></div>',
  };
  const GlLoadingIcon = { name: 'gl-loading-icon', template: '<div>loading</div>' };

  const findEmptyState = () => wrapper.findComponent(GlEmptyState);
  const findListComponent = () => wrapper.findComponent(PackageList);
  const findInfrastructureSearch = () => wrapper.findComponent(InfrastructureSearch);
  const findInfrastructureTitle = () => wrapper.findComponent(InfrastructureTitle);

  const createStore = ({ filter = [], packageCount = 0 } = {}) => {
    store = new Vuex.Store({
      state: {
        isLoading: false,
        filter,
        pagination: {
          total: packageCount,
        },
      },
    });
    store.dispatch = jest.fn();
  };

  const mountComponent = ({ isGroupPage = false } = {}) => {
    wrapper = shallowMount(PackageListApp, {
      store,
      provide: {
        isGroupPage,
        emptyListIllustration: 'helpSvg',
        resourceId: 'project_id',
      },
      stubs: {
        GlEmptyState,
        GlLoadingIcon,
        PackageList,
        GlSprintf,
        GlLink,
      },
    });
  };

  beforeEach(() => {
    createStore();
    jest.spyOn(packageUtils, 'getQueryParams').mockReturnValue({});
    mountComponent();
  });

  it('renders', () => {
    createStore({ packageCount: 1 });
    mountComponent();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('calls requestPackagesList on page:changed', () => {
    const list = findListComponent();
    list.vm.$emit('page:changed', 1);
    expect(store.dispatch).toHaveBeenCalledWith('requestPackagesList', {
      page: 1,
      isGroupPage: false,
      resourceId: 'project_id',
    });
  });

  it('calls requestDeletePackage on package:delete', () => {
    const payload = {
      _links: {
        delete_api_path: 'foo',
      },
    };
    const list = findListComponent();
    list.vm.$emit('package:delete', payload);

    expect(store.dispatch).toHaveBeenCalledWith('requestDeletePackage', {
      ...payload,
      isGroupPage: false,
      resourceId: 'project_id',
    });
  });

  it('calls requestPackagesList only once on render', () => {
    expect(store.dispatch).toHaveBeenCalledTimes(3);
    expect(store.dispatch).toHaveBeenNthCalledWith(1, 'setSorting', expect.any(Object));
    expect(store.dispatch).toHaveBeenNthCalledWith(2, 'setFilter', expect.any(Array));
    expect(store.dispatch).toHaveBeenNthCalledWith(3, 'requestPackagesList', {
      isGroupPage: false,
      resourceId: 'project_id',
    });
  });

  describe('url query string handling', () => {
    const defaultQueryParamsMock = {
      search: [1, 2],
      type: 'npm',
      sort: 'asc',
      orderBy: 'created',
    };

    beforeEach(() => {
      createStore();
      jest.spyOn(packageUtils, 'getQueryParams').mockReturnValue(defaultQueryParamsMock);
    });

    it('calls setSorting with the query string based sorting', () => {
      mountComponent();

      expect(store.dispatch).toHaveBeenNthCalledWith(1, 'setSorting', {
        orderBy: defaultQueryParamsMock.orderBy,
        sort: defaultQueryParamsMock.sort,
      });
    });

    it('calls setFilter with the query string based filters', () => {
      mountComponent();

      expect(store.dispatch).toHaveBeenNthCalledWith(2, 'setFilter', [
        { type: 'type', value: { data: defaultQueryParamsMock.type } },
        { type: FILTERED_SEARCH_TERM, value: { data: defaultQueryParamsMock.search[0] } },
        { type: FILTERED_SEARCH_TERM, value: { data: defaultQueryParamsMock.search[1] } },
      ]);
    });

    it('calls setSorting and setFilters with the results of extractFilterAndSorting', () => {
      jest
        .spyOn(packageUtils, 'extractFilterAndSorting')
        .mockReturnValue({ filters: ['foo'], sorting: { sort: 'desc' } });

      mountComponent();

      expect(store.dispatch).toHaveBeenNthCalledWith(1, 'setSorting', { sort: 'desc' });
      expect(store.dispatch).toHaveBeenNthCalledWith(2, 'setFilter', ['foo']);
    });
  });

  describe('empty state', () => {
    const heading = () => findEmptyState().find('h2');

    it('generate the correct empty list link', () => {
      const link = findListComponent().findComponent(GlLink);

      expect(link.attributes('href')).toBe(
        helpPagePath('user/packages/terraform_module_registry/_index'),
      );
      expect(link.text()).toBe('publish and share your packages');
    });

    it('includes the right content on the default tab', () => {
      expect(heading().text()).toBe('You have no Terraform modules in your project');
    });

    it('does not show infrastructure registry title', () => {
      expect(findInfrastructureTitle().exists()).toBe(false);
    });

    describe('when group page', () => {
      beforeEach(() => {
        createStore();
        mountComponent({ isGroupPage: true });
      });

      it('includes the right content', () => {
        expect(heading().text()).toBe('You have no Terraform modules in your group');
      });
    });
  });

  describe('filter without results', () => {
    beforeEach(() => {
      createStore({ filter: [{ type: 'something' }] });
      mountComponent();
    });

    it('should show specific empty message', () => {
      expect(findEmptyState().text()).toContain('Sorry, your filter produced no results');
      expect(findEmptyState().text()).toContain(
        'To widen your search, change or remove the filters above',
      );
    });
  });

  describe('search', () => {
    describe('with no packages', () => {
      it('does not exist', () => {
        expect(findInfrastructureSearch().exists()).toBe(false);
      });
    });

    describe('with packages', () => {
      beforeEach(() => {
        createStore({ packageCount: 1 });
        mountComponent();
      });

      it('exists', () => {
        expect(findInfrastructureSearch().exists()).toBe(true);
      });

      it('shows infrastructure registry title', () => {
        expect(findInfrastructureTitle().exists()).toBe(true);
      });

      it('on update fetches data from the store', () => {
        store.dispatch.mockClear();

        findInfrastructureSearch().vm.$emit('update');

        expect(store.dispatch).toHaveBeenCalledWith('requestPackagesList', {
          isGroupPage: false,
          resourceId: 'project_id',
        });
      });
    });
  });

  describe('delete alert handling', () => {
    const originalLocation = window.location.href;
    const search = `?${SHOW_DELETE_SUCCESS_ALERT}=true`;

    beforeEach(() => {
      createStore();
      jest.spyOn(commonUtils, 'historyReplaceState').mockImplementation(() => {});
      setWindowLocation(search);
    });

    afterEach(() => {
      setWindowLocation(originalLocation);
    });

    it(`creates an alert if the query string contains ${SHOW_DELETE_SUCCESS_ALERT}`, () => {
      mountComponent();

      expect(createAlert).toHaveBeenCalledWith({
        message: DELETE_PACKAGE_SUCCESS_MESSAGE,
        variant: VARIANT_INFO,
      });
    });

    it('calls historyReplaceState with a clean url', () => {
      mountComponent();

      expect(commonUtils.historyReplaceState).toHaveBeenCalledWith(originalLocation);
    });

    it(`does nothing if the query string does not contain ${SHOW_DELETE_SUCCESS_ALERT}`, () => {
      setWindowLocation('?');
      mountComponent();

      expect(createAlert).not.toHaveBeenCalled();
      expect(commonUtils.historyReplaceState).not.toHaveBeenCalled();
    });
  });
});
