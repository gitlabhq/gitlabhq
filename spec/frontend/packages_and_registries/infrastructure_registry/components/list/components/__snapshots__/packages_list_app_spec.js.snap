// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages_list_app renders 1`] = `
<div>
  <infrastructure-title-stub
    count="1"
    helpurl="/help/user/packages/terraform_module_registry/_index"
  />
  <infrastructure-search-stub />
  <div>
    <section
      class="gl-empty-state gl-flex gl-flex-col gl-text-center"
    >
      <div
        class="gl-max-w-full"
      >
        <img
          alt=""
          class="gl-dark-invert-keep-hue gl-max-w-full"
          height="144"
          src="helpSvg"
        />
      </div>
      <div
        class="gl-empty-state-content gl-m-auto gl-mx-auto gl-my-0 gl-p-5"
        data-testid="gl-empty-state-content"
      >
        <h2
          class="gl-leading-36 gl-mb-0 gl-mt-0 gl-text-size-h-display h4"
        >
          You have no Terraform modules in your project
        </h2>
        <p
          class="gl-mb-0 gl-mt-4 gl-text-subtle"
        >
          Learn how to
          <a
            class="gl-link"
            href="/help/user/packages/terraform_module_registry/_index"
            rel="noopener noreferrer"
            target="_blank"
          >
            publish and share your packages
          </a>
          with GitLab.
        </p>
        <div
          class="gl-flex gl-flex-wrap gl-justify-center gl-mt-5"
        />
      </div>
    </section>
  </div>
</div>
`;
